sdk/program/src/stable_layout/stable_instruction.rs

```diff
@@ -0,0 +1,77 @@
+//! Instruction, with a stable memory layout
+
+use {
+    crate::{
+        instruction::{AccountMeta, Instruction},
+        pubkey::Pubkey,
+        stable_layout::stable_vec::StableVec,
+    },
+    std::fmt::Debug,
+};
+
+#[derive(Debug)]
+#[repr(C)]
+pub struct StableInstruction {
+    pub accounts: StableVec<AccountMeta>,
+    pub data: StableVec<u8>,
+    pub program_id: Pubkey,
+}
+
+impl From<Instruction> for StableInstruction {
+    fn from(other: Instruction) -> Self {
+        Self {
+            accounts: StableVec::from(other.accounts),
+            data: StableVec::from(other.data),
+            program_id: other.program_id,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use {
+        super::*,
+        memoffset::offset_of,
+        std::mem::{align_of, size_of},
+    };
+
+    #[test]
+    fn test_memory_layout() {
+        assert_eq!(offset_of!(StableInstruction, accounts), 0);
+        assert_eq!(offset_of!(StableInstruction, data), 24);
+        assert_eq!(offset_of!(StableInstruction, program_id), 48);
+        assert_eq!(align_of::<StableInstruction>(), 8);
+        assert_eq!(size_of::<StableInstruction>(), 24 + 24 + 32);
+
+        let program_id = Pubkey::new_unique();
+        let account_meta1 = AccountMeta {
+            pubkey: Pubkey::new_unique(),
+            is_signer: true,
+            is_writable: false,
+        };
+        let account_meta2 = AccountMeta {
+            pubkey: Pubkey::new_unique(),
+            is_signer: false,
+            is_writable: true,
+        };
+        let accounts = vec![account_meta1, account_meta2];
+        let data = vec![1, 2, 3, 4, 5];
+        let instruction = Instruction {
+            program_id,
+            accounts: accounts.clone(),
+            data: data.clone(),
+        };
+        let instruction = StableInstruction::from(instruction);
+
+        let instruction_addr = &instruction as *const _ as u64;
+
+        let accounts_ptr = instruction_addr as *const &[AccountMeta; 2];
+        assert_eq!(unsafe { *accounts_ptr }, accounts.as_slice());
+
+        let data_ptr = (instruction_addr + 24) as *const &[u8; 5];
```

Comment:
Same here, I think you're getting lucky that len == cap (vec![] sets capacity to
the n of elements passed to the macro), so it accidentally matches the slice